v1 Commit 6600bd7 causing goagen to panic

Introduced with commit 6600bd7

```
panic: runtime error: comparing uncomparable type design.Object

goroutine 1 [running]:
xxxx/vendor/github.com/goadesign/goa/design.hasFile(0x8dfa60, 0xc4202b23f0, 0xc42040a8d0, 0x1, 0x1, 0xc42040a8d0)
	/home/msonghurst/go/src/xxxx/vendor/github.com/goadesign/goa/design/types.go:613 +0x91
xxxx/vendor/github.com/goadesign/goa/design.hasFile(0x8dfa60, 0xc42023f830, 0xc42040a8d0, 0x1, 0x1, 0xc42028f590)
	/home/msonghurst/go/src/xxxx/vendor/github.com/goadesign/goa/design/types.go:635 +0x416
xxxx/vendor/github.com/goadesign/goa/design.HasFile(0x8dfa60, 0xc42023f830, 0xe)
	/home/msonghurst/go/src/xxxx/vendor/github.com/goadesign/goa/design/types.go:605 +0x46
xxxx/vendor/github.com/goadesign/goa/design.(*ActionDefinition).Validate(0xc42024e5a0, 0xc4202fcde0)
	/home/msonghurst/go/src/xxxx/vendor/github.com/goadesign/goa/design/validation.go:346 +0xe76
xxxx/vendor/github.com/goadesign/goa/design.(*ResourceDefinition).validateActions(0xc420292380, 0xc42042ac00)
	/home/msonghurst/go/src/xxxx/vendor/github.com/goadesign/goa/design/validation.go:240 +0xf2
xxxx/vendor/github.com/goadesign/goa/design.(*ResourceDefinition).Validate(0xc420292380, 0xc42028f948)
	/home/msonghurst/go/src/xxxx/vendor/github.com/goadesign/goa/design/validation.go:218 +0x6f
xxxx/vendor/github.com/goadesign/goa/design.(*APIDefinition).Validate.func1(0xc420292380, 0x0, 0x0)
	/home/msonghurst/go/src/xxxx/vendor/github.com/goadesign/goa/design/validation.go:91 +0x55
xxxx/vendor/github.com/goadesign/goa/design.(*APIDefinition).IterateResources(0xc42021e420, 0xc42028fbb8, 0xc4201c1c00, 0x820201)
	/home/msonghurst/go/src/xxxx/vendor/github.com/goadesign/goa/design/definitions.go:647 +0x219
xxxx/vendor/github.com/goadesign/goa/design.(*APIDefinition).Validate(0xc42021e420, 0x8d66a0, 0xc42021e420)
	/home/msonghurst/go/src/xxxx/vendor/github.com/goadesign/goa/design/validation.go:90 +0x1ee
xxxx/vendor/github.com/goadesign/goa/dslengine.validateSet(0xc42040a3c0, 0x1, 0x1, 0xb, 0xc)
	/home/msonghurst/go/src/xxxx/vendor/github.com/goadesign/goa/dslengine/runner.go:314 +0xdb
xxxx/vendor/github.com/goadesign/goa/design.(*APIDefinition).IterateSets(0xc42021e420, 0x8a1878)
	/home/msonghurst/go/src/xxxx/vendor/github.com/goadesign/goa/design/definitions.go:461 +0x92
xxxx/vendor/github.com/goadesign/goa/dslengine.Run(0x8d71a0, 0xb30820)
	/home/msonghurst/go/src/xxxx/vendor/github.com/goadesign/goa/dslengine/runner.go:108 +0x1d7
main.main()
	/home/msonghurst/go/src/xxxx/goagen546751901/main.go:24 +0x5a
```

--- a/design/types.go
+++ b/design/types.go
@@ -609,13 +609,6 @@ func hasFile(dt DataType, traversed []DataType) bool {
 	if dt == nil {
 		return false
 	}
-	for _, v := range traversed {
-		if dt == v {
-			return false
-		}
-	}
-	traversed = append(traversed, dt)
-
 	switch {
 	case dt.IsPrimitive():
 		return dt.Kind() == FileKind
@@ -630,6 +623,14 @@ func hasFile(dt DataType, traversed []DataType) bool {
 		if hasFile(dt.ToHash().ElemType.Type, traversed) {
 			return true
 		}
+	case dt.IsObject() && dt.Kind() == UserTypeKind:
+		for _, v := range traversed {
+			if dt == v {
+				return false
+			}
+		}
+		traversed = append(traversed, dt)
+		fallthrough
 	case dt.IsObject():
 		for _, att := range dt.ToObject() {
 			if hasFile(att.Type, traversed) {

--- a/goagen/gen_swagger/swagger_test.go
+++ b/goagen/gen_swagger/swagger_test.go
@@ -334,6 +334,12 @@ var _ = Describe("New", func() {
 			BeforeEach(func() {
 				p := Type("RecursivePayload", func() {
 					Member("m1", "RecursivePayload")
+					Member("m2", ArrayOf("RecursivePayload"))
+					Member("m3", HashOf(String, "RecursivePayload"))
+					Member("m4", func() {
+						Member("m5", String)
+						Member("m6", "RecursivePayload")
+					})
 				})
 				Resource("res", func() {
 					Action("act", func() {

